File: backend/app/schemas/course.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.video import VideoResponse


class CourseCreate(BaseModel):
    playlist_url: str


class CourseSeed(BaseModel):
    title: str
    playlist_url: str
    playlist_id: str


class CourseResponse(BaseModel):
    id: UUID
    title: str
    playlist_url: str
    playlist_id: str
    video_count: int
    status: str
    created_at: datetime | None
    updated_at: datetime | None

    model_config = ConfigDict(from_attributes=True)


class CourseDetail(CourseResponse):
    videos: list[VideoResponse]


class CourseStatusResponse(BaseModel):
    course_id: UUID
    total: int
    pending: int
    processing: int
    completed: int
    failed: int
    deferred: int
    deferred_until: datetime | None = None
    quota_remaining: dict[str, int] = Field(default_factory=dict)

File: frontend/src/types/course.ts
import type { VideoResponse } from "./video";

export interface CourseCreate {
  playlist_url: string;
}

export interface CourseResponse {
  id: string;
  title: string;
  playlist_url: string;
  playlist_id: string;
  video_count: number;
  status: "pending" | "processing" | "completed" | "partial";
  created_at: string | null;
  updated_at: string | null;
}

export interface CourseDetail extends CourseResponse {
  videos: VideoResponse[];
}

export interface CourseStatusResponse {
  course_id: string;
  total: number;
  pending: number;
  processing: number;
  rate_limited: number;
  batch_processing: number;
  completed: number;
  failed: number;
  deferred: number;
  deferred_until: string | null;
  next_retry_at: string | null;
  quota_remaining: {
    llm_requests?: number;
    llm_tokens?: number;
    whisper_requests?: number;
    whisper_audio_seconds_hour?: number;
    whisper_audio_seconds_day?: number;
  };
}
